docs/Toy/toy/dialect.py

```diff
@@ -0,0 +1,267 @@
+                            cast(TensorType[Any], lhs.typ).element_type)
+        return super().create(result_types=[element_type], operands=[lhs, rhs])
+
+
+@irdl_op_definition
+class FuncOp(Operation):
+    """
+    The "toy.func" operation represents a user defined function. These are
+    callable SSA-region operations that contain toy computations.
+
+    Example:
+
+    ```mlir
+    toy.func @main() {
+      %0 = toy.constant dense<5.500000e+00> : tensor<f64>
+      %1 = toy.reshape(%0 : tensor<f64>) to tensor<2x2xf64>
+      toy.print %1 : tensor<2x2xf64>
+      toy.return
+    }
+    ```
+    """
+    name = 'toy.func'
+    body: Region
+    sym_name: OpAttr[StringAttr]
+    function_type: OpAttr[FunctionType]
+    sym_visibility: OptOpAttr[StringAttr]
+
+    @staticmethod
+    def from_region(name: str, ftype: FunctionType, region: Region):
+        return FuncOp.create(attributes={
+            "sym_name": StringAttr.from_str(name),
+            "function_type": ftype,
+            "sym_visibility": StringAttr.from_str("private"),
+        },
+                             regions=[region])
+
+    @staticmethod
+    def from_callable(name: str,
+                      input_types: List[Attribute],
+                      return_types: List[Attribute],
+                      func: Block.BlockCallback,
+                      private: bool = False):
+        type_attr = FunctionType.from_lists(input_types, return_types)
+        attributes = {
+            "sym_name": name,
+            "function_type": type_attr,
+        }
+        if private:
+            attributes["sym_visibility"] = "private"
+        return FuncOp.build(attributes=attributes,
+                            regions=[
+                                Region.from_block_list(
+                                    [Block.from_callable(input_types, func)])
+                            ])
+
+
+@irdl_op_definition
+class GenericCallOp(Operation):
+    name: str = "toy.generic_call"
+    arguments: Annotated[VarOperand, AnyAttr()]
+    callee: OpAttr[FlatSymbolRefAttr]
+
+    # Note: naming this results triggers an ArgumentError
+    res: Annotated[VarOpResult, AnyTensorTypeF64]
+
+    @classmethod
+    def get(cls: type[GenericCallOp], callee: Union[str, FlatSymbolRefAttr],
+            operands: List[Union[SSAValue, OpResult]],
+            return_types: List[Attribute]) -> GenericCallOp:
+        if isinstance(callee, str):
+            callee = FlatSymbolRefAttr.from_str(callee)
+
+        return super().create(operands=operands,
+                              result_types=return_types,
+                              attributes={"callee": callee})
+
+
+@irdl_op_definition
+class MulOp(Operation):
+    """
+    The "mul" operation performs element-wise multiplication between two
+    tensors. The shapes of the tensor operands are expected to match.
+    """
+    name: str = 'toy.mul'
+    arguments: Annotated[VarOperand, AnyTensorTypeF64]
```

Comment:
Weirdly the MLIR tutorial defines these as variadic, as far as I understand, although it could well have been `lhs` and `rhs` instead, keeping as-is.

Reply:
I think a variadic tensor multiplication makes sense, given that choosing the order of those multiplications can have a big impact on the cost, having them nicely packed is nice to reason about it and is appropriate at that specific abstraction level, in my opinion.

Reply:
Not sure what you mean :)

Reply:
To give my two cents: variadic multiplies are usually a way to wrap things like `a * b * c` into one multiplication (with 3 operands). This has the big advantage that you can more easily define the order, in contrast to having either `(a * b) * c` or `a * (b * c)` in the case of binary multiplies, which encodes the order in which the multiplications will be executed.

I agree with Emilie. On some abstraction level, you do not care about order, but lower down you do, as it can matter for performance.

Reply:
This would be true if they weren't pointwise multiplies :)

Reply:
Oh I missed the point-wise, so the performance argument doesn't hold for this specific op.
It does on many tensor ops though, so I feel like it is a sensible design.
A dialect is also designed to reason on a specific abstraction; I feel like a variadic product maps well with an algebraic abstraction and reasoning, when one often reasons about the product as a whole instead of reasoning about a chain of products. As in $\prod_{1\le i \le n}{T_i}$ instead of $T_1\cdot T_2 \cdot \ldots \cdot T_n$

By doing so, you can put the product associativity to use when rewriting a combination of long products with other operations, which would be a bit more painful if you'd have to do so on sequences of binary ops every time.